Editing an attachment in mail crashes after adding release assert in 292439@main.

megangardner opened the pull request:
#### 0ffe3e1de9f2dc1b37a1cd7874ba30751800e55d
<pre>
Editing an attachment in mail crashes after adding release assert in 292439@main.
<a href="https://bugs.webkit.org/show_bug.cgi?id=291719">https://bugs.webkit.org/show_bug.cgi?id=291719</a>
<a href="https://rdar.apple.com/149029883">rdar://149029883</a>

Reviewed by Chris Dumez and Abrar Rahman Protyasha.

This work should have been done on the main thread this whole time, as exposed
by adding the release assert in <a href="https://commits.webkit.org/292439@main">https://commits.webkit.org/292439@main</a>
So to fix we need to move the work to the main thread.

* Source/WebKit/UIProcess/mac/WKSharingServicePickerDelegate.mm:
(-[WKSharingServicePickerDelegate sharingService:didShareItems:]):

Canonical link: <a href="https://commits.webkit.org/293875@main">https://commits.webkit.org/293875@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/acc8767bfa2c93e8f49bd2a37d78de85f927fd7f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100111 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19759 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105241 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50692 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20065 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28232 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76207 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33276 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103118 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15336 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90444 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56568 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15151 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50062 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85063 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8525 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107600 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27224 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19945 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85162 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27587 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86651 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84697 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21531 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29373 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7133 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21079 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27161 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32393 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26972 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30288 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28531 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/mac/WKSharingServicePickerDelegate.mm:
```diff
@@ -151,21 +151,23 @@ - (void)sharingService:(NSSharingService *)sharingService didShareItems:(NSArray
-            RetainPtr attachment = wrapper(apiAttachment);
-            [attachment setData:data newContentType:itemUTI.get()];
-            webPage->didInvalidateDataForAttachment(*apiAttachment.get());
+            ensureOnMainRunLoop([weakPage = WTFMove(weakPage), attachmentID, itemUTI = WTFMove(itemUTI), data = retainPtr(data), error = retainPtr(error)] {
```
`itemUTI = WTFMove(itemUTI)` this doesn't actually move, so we can just not move

aprotyas replied:
Also, I think our static analysis tooling prefers direct `RetainPtr` construction (rather than using the `retainPtr(...)` helper method.

aprotyas replied:
In all...
```suggestion
            ensureOnMainRunLoop([weakPage = WTFMove(weakPage), attachmentID, itemUTI, data = RetainPtr { data }, error = RetainPtr { error }] {
```

cdumez commented on Source/WebKit/UIProcess/mac/WKSharingServicePickerDelegate.mm:
```diff
@@ -151,21 +151,23 @@ - (void)sharingService:(NSSharingService *)sharingService didShareItems:(NSArray
+
+                RetainPtr attachment = wrapper(apiAttachment.get());
+                [attachment setData:data.get() newContentType:itemUTI.get()];
+                webPage->didInvalidateDataForAttachment(*apiAttachment.get());
```
nit: `.get()` is unnecessary.

cdumez commented on Source/WebKit/UIProcess/mac/WKSharingServicePickerDelegate.mm:
```diff
@@ -151,21 +151,23 @@ - (void)sharingService:(NSSharingService *)sharingService didShareItems:(NSArray
+                if (!webPage)
+                    return;
+
+                if (error)
```
Seems a little unfortunate to dispatch to the background thread just to early return. I recommend we null check the error on the background thread and early return there. Then we don't need to capture `error` in the lambda.

megangardner committed:
Editing an attachment in mail crashes after adding release assert in 292439@main.
https://bugs.webkit.org/show_bug.cgi?id=291719
rdar://149029883

Reviewed by Chris Dumez and Abrar Rahman Protyasha.

This work should have been done on the main thread this whole time, as exposed
by adding the release assert in https://commits.webkit.org/292439@main
So to fix we need to move the work to the main thread.

* Source/WebKit/UIProcess/mac/WKSharingServicePickerDelegate.mm:
(-[WKSharingServicePickerDelegate sharingService:didShareItems:]):

Canonical link: https://commits.webkit.org/293875@main
